File: python/queue.py
# A `queue` is a data structure that follows the FIFO (First In, First Out)
# principle. Elements are added to the end of the queue (enqueue) and removed
# from the front (dequeue). A queue is implemented using a linked list as the
# underlying data structure for storing its elements

# Enqueue updates the tail by adding a new element, and dequeue pops the head
# by removing the element at the front


class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.next: Node | None = None


class Queue:
    def __init__(self) -> None:
        self.__head: Node | None = None
        self.__tail: Node | None = None
        self.length: int = 0

    def enqueue(self, data) -> None:
        new_node = Node(data)

        if not self.__tail:
            self.__tail = new_node
            self.__head = new_node

            self.length += 1
            return

        self.__tail.next = new_node
        self.__tail = new_node

        self.length += 1

    # Returns data of removed node or None
    def dequeue(self):
        if not self.__head:
            return

        current_head = self.__head
        self.__head = self.__head.next

        self.length -= 1

        if self.length == 0:
            self.__tail = None
            self.__head = None

        return current_head.data

    # Returns data of top node (head) or None
    def peak(self):
        return self.__head.data if self.__head else None

    def __str__(self) -> str:
        msg = f"Length {self.length}: "
        current_node = self.__head

        while current_node:
            msg += f"|{current_node.data}| -> "
            current_node = current_node.next

        msg += "END"
        return msg


queue = Queue()

print(queue)

queue.enqueue(3)

print(queue)

queue.enqueue(4)

print(queue)

queue.dequeue()

print(queue)

print(queue.peak())

File: typescript/queue.ts
import { strict as assert } from "node:assert";

type TNode<T> = {
    data: T;
    next?: TNode<T>;
};

class Queue<T> {
    public length: number;
    private head: TNode<T> | undefined;
    private tail: TNode<T> | undefined;

    constructor() {
        this.length = 0;
        this.head = this.tail = undefined;
    }

    enqueue(data: T): void {
        const node = { data: data } as TNode<T>;

        if (this.length === 0) {
            this.head = this.tail = node;
            this.length++;
            return;
        }

        this.length++;

        this.tail!.next = node;
        this.tail = node;
    }

    deque(): T | undefined {
        if (this.length === 0) return;

        this.length = Math.max(0, --this.length);

        const head = this.head;

        if (this.length === 0) {
            this.head = this.tail = undefined;
            delete head?.next;
            return head?.data;
        }

        this.head = this.head?.next;
        delete head?.next;
        return head?.data;
    }

    peek(): T | undefined {
        return this.head?.data;
    }
}

const queue = new Queue<number>();

queue.enqueue(5);
queue.enqueue(7);
queue.enqueue(9);

assert(queue.deque() === 5, "Expected 5 to be dequeued");
assert(queue.length === 2, "Expected queue length to be 2");

queue.enqueue(11);
assert(queue.deque() === 7, "Expected 7 to be dequeued");
assert(queue.deque() === 9, "Expected 9 to be dequeued");
assert(queue.peek() === 11, "Expected 11 to be at the front of the queue");
assert(queue.deque() === 11, "Expected 11 to be dequeued");
assert(
    queue.deque() === undefined,
    "Expected dequeuing an empty queue to return undefined",
);
assert(queue.length === 0, "Expected queue length to be 0");

queue.enqueue(69);
assert(queue.peek() === 69, "Expected 69 to be at the front of the queue");
assert(queue.length === 1, "Expected queue length to be 1");

console.log("ALL TESTS PASSED!");
